docarray/typing/url/audio_url.py

```diff
@@ -0,0 +1,114 @@
+import wave
+from typing import TYPE_CHECKING, Any, Type, TypeVar, Union
+
+import numpy as np
+
+from docarray.typing.url.any_url import AnyUrl
+
+if TYPE_CHECKING:
+    from pydantic import BaseConfig
+    from pydantic.fields import ModelField
+
+    from docarray.proto import NodeProto
+
+T = TypeVar('T', bound='AudioUrl')
+
+AUDIO_FILE_FORMATS = ['wav']
+
+
+class AudioUrl(AnyUrl):
+    """
+    URL to a .wav file.
+    Can be remote (web) URL, or a local file path.
+    """
+
+    def _to_node_protobuf(self: T) -> 'NodeProto':
+        """Convert Document into a NodeProto protobuf message. This function should
+        be called when the Document is nested into another Document that needs to
+        be converted into a protobuf
+
+        :return: the nested item protobuf message
+        """
+        from docarray.proto import NodeProto
+
+        return NodeProto(audio_url=str(self))
+
+    @classmethod
+    def validate(
+        cls: Type[T],
+        value: Union[T, np.ndarray, Any],
+        field: 'ModelField',
+        config: 'BaseConfig',
+    ) -> T:
+        url = super().validate(value, field, config)  # basic url validation
+        has_audio_extension = any(url.endswith(ext) for ext in AUDIO_FILE_FORMATS)
+        if not has_audio_extension:
+            raise ValueError(
+                f'Audio URL must have one of the following extensions:'
+                f'{AUDIO_FILE_FORMATS}'
+            )
+        return cls(str(url), scheme=None)
+
+    def load(self: T) -> np.ndarray:
+        """
+        Load the data from the url into a numpy.ndarray.
+
+        EXAMPLE USAGE
+
+        .. code-block:: python
+
+            from docarray import Document
+            import numpy as np
+
+            from docarray.typing import AudioUrl
+
+
+            class MyDoc(Document):
+                audio_url: AudioUrl
+
+
+            doc = MyDoc(audio_url="toydata/hello.wav")
+
+            audio_tensor = doc.audio_url.load()
+            assert isinstance(audio_tensor, np.ndarray)
+
+        :return: np.ndarray representing the audio file content
+        """
+        import io
+
+        file: Union[io.BytesIO, T]
+
+        if self.startswith('http'):
```

Comment:
how is this `startswith` implemented on `self`? how is `T` guaranteed to implement this method?